packages/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -358,22 +371,134 @@ export class AmqplibInstrumentation extends InstrumentationBase<typeof amqp> {
+        ): boolean {
+            if (isConfirmChannel(context.active())) {
+                // work already done
+                return original.apply(this, arguments);
+            } else {
+                const channel = this;
+                const { span, modifiedOptions } = self.createPublishSpan(
+                    self,
+                    exchange,
+                    routingKey,
+                    channel,
+                    moduleVersion,
+                    options
+                );
+
+                if (self._config.publishHook) {
+                    safeExecuteInTheMiddle(
+                        () => self._config.publishHook(span, { exchange, routingKey, content, options }, false),
+                        (e) => {
+                            if (e) {
+                                diag.error('amqplib instrumentation: publishHook error', e);
+                            }
+                        },
+                        true
+                    );
+                }
+
+                // calling normal channel publish function is only storing the message in queue.
+                // it does not send it and waits for an ack, so the span duration is expected to be very short.
+                const argumentsCopy = [...arguments];
+                argumentsCopy[3] = modifiedOptions;
+                const originalRes = original.apply(this, argumentsCopy);
+                span.end();
+                return originalRes;
             }
-            span.end();
-            return originalRes;
         };
     }
 
+    private createPublishSpan(
+        self: this,
+        exchange: string,
+        routingKey: string,
+        channel,
+        moduleVersion: string,
+        options?: amqp.Options.Publish,
+        isConfirmChannel?: boolean
+    ) {
+        const normalizedExchange = normalizeExchange(exchange);
+
+        const span = self.tracer.startSpan(`${normalizedExchange} -> ${routingKey} send`, {
+            kind: SpanKind.PRODUCER,
+            attributes: {
+                ...channel.connection[CONNECTION_ATTRIBUTES],
+                [SemanticAttributes.MESSAGING_DESTINATION]: exchange,
+                [SemanticAttributes.MESSAGING_DESTINATION_KIND]: MessagingDestinationKindValues.TOPIC,
+                [SemanticAttributes.MESSAGING_RABBITMQ_ROUTING_KEY]: routingKey,
+                [SemanticAttributes.MESSAGING_MESSAGE_ID]: options?.messageId,
+                [SemanticAttributes.MESSAGING_CONVERSATION_ID]: options?.correlationId,
+            },
+        });
+        if (self._config.moduleVersionAttributeName) {
+            span.setAttribute(self._config.moduleVersionAttributeName, moduleVersion);
+        }
+        const modifiedOptions = options ?? {};
+        modifiedOptions.headers = modifiedOptions.headers ?? {};
+
+        let markedContext = context.active();
+        if (isConfirmChannel) {
+            markedContext = markedContext.setValue(IS_CONFIRM_CHANNEL, true);
```

Comment:
Okay. I like it, I have seen it and actually don't like mixing both key/method in instrumentation file. It will be cleaner this way. I put this in utils, not sure it should be moved to another file?

Reply:
I think utils is a good place. thanks for adding it :)